upup/pkg/fi/cloudup/apply_cluster.go

```diff
@@ -1410,20 +1410,32 @@ func (n *nodeUpConfigBuilder) BuildConfig(ig *kops.InstanceGroup, apiserverAddit
 
 	// Set API server address to an IP from the cluster network CIDR
 	if cluster.UsesNoneDNS() {
-		for _, networkCIDR := range append(cluster.Spec.Networking.AdditionalNetworkCIDRs, cluster.Spec.Networking.NetworkCIDR) {
-			_, cidr, err := net.ParseCIDR(networkCIDR)
-			if err != nil {
-				return nil, nil, fmt.Errorf("failed to parse network CIDR %q: %w", networkCIDR, err)
-			}
-			for _, additionalIP := range apiserverAdditionalIPs {
-				if cidr.Contains(net.ParseIP(additionalIP)) {
-					bootConfig.APIServerIP = additionalIP
+		switch cluster.Spec.GetCloudProvider() {
+		case kops.CloudProviderAWS, kops.CloudProviderHetzner:
+			// Use a private IP address that belongs to the cluster network CIDR (some additional addresses may be FQDNs or public IPs)
+			for _, networkCIDR := range append(cluster.Spec.Networking.AdditionalNetworkCIDRs, cluster.Spec.Networking.NetworkCIDR) {
+				_, cidr, err := net.ParseCIDR(networkCIDR)
+				if err != nil {
+					return nil, nil, fmt.Errorf("failed to parse network CIDR %q: %w", networkCIDR, err)
+				}
+				for _, additionalIP := range apiserverAdditionalIPs {
+					if cidr.Contains(net.ParseIP(additionalIP)) {
+						bootConfig.APIServerIP = additionalIP
+						break
+					}
+				}
+				if bootConfig.APIServerIP != "" {
 					break
 				}
 			}
-			if bootConfig.APIServerIP != "" {
+		case kops.CloudProviderGCE:
+			// Use any IP address that is found (including public ones)
```

Comment:
Nit (for a future PR): We probably should just have nodeup accept a list of IPs, probably easier than choosing!

Reply:
I thought about this a little. Would be nice to do the choosing on nodeup side. For sure it would be better equipped to find the best bootstrap address for itself.